action/protocol/staking/candidate_statereader.go

```diff
@@ -110,6 +112,41 @@ func (c *candSR) AllCandidates() CandidateList {
 	return c.view.candCenter.All()
 }
 
+func (c *candSR) IsActiveCandidateAt(cand *Candidate, height uint64) (bool, error) {
+	// self-stake bucket should be either self-owned or endorsed
+	vb, err := c.getBucket(cand.SelfStakeBucketIdx)
+	if err != nil {
+		return false, errors.Wrapf(err, "failed to get bucket %d", cand.SelfStakeBucketIdx)
+	}
+	// self-owned bucket should not be unstaked
+	if address.Equal(vb.Owner, cand.Owner) {
+		return !vb.isUnstaked(), nil
+	}
+	// endorsed bucket should not be expired
+	if !address.Equal(vb.Candidate, cand.Owner) {
+		return false, nil
+	}
+	esr := NewEndorsementStateReader(c.SR())
+	endorse, err := esr.Get(cand.SelfStakeBucketIdx)
```

Comment:
this should be guarded by hard-fork flag? before 1.14, no such thing as endorsement exist

Reply:
hard-fork at https://github.com/iotexproject/iotex-core/pull/4062/files#diff-6bac7263ac5bdd5a72fb2bbb66bff76ebd0b794b0f998ed62bfafab7bba319e6R479